.github/workflows/image-builder.yml

```diff
@@ -0,0 +1,181 @@
+        with:
+          name: ${{ matrix.image }}-${{ matrix.os }}-tarball
+          path: ${{ runner.temp }}/${{ matrix.image }}-${{ matrix.os }}.tar
+          if-no-files-found: error
+          retention-days: 1
+
+  build-images:
+    needs: [build-base-images]
+    strategy:
+      max-parallel: 5
+      matrix:
+        image: ${{ fromJSON(inputs.images) }}
+        os: [ubuntu-24.04, ubuntu-24.04-arm]
+        exclude: ${{ fromJSON(inputs.exclude) }}
+    runs-on: ${{ matrix.os }}
+    env:
+      CAN_PUSH_IMAGE: ${{ secrets.DOCKERHUB_USERNAME != '' && github.ref_name == github.event.repository.default_branch }}
+
+    # Associate each workspace with requirements
+    steps:
+      # Checkout the repository with full depth so we can check if the workspace was modified
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Check whether ${{ inputs.folder }}/${{ matrix.image }} was modified
+        id: check-modified
+        run: |
+          scripts/check_path_modified.sh ${{ inputs.folder }}/${{ matrix.image }} ${{ inputs.folder }}_${{ matrix.image }}_modified
+
+      - name: Set up Docker Buildx
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: docker/setup-buildx-action@v3
+
+      - name: Download artifact
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: actions/download-artifact@v4
+        # Download all uploaded artifacts that match our OS
+        with:
+          pattern: '*-${{ matrix.os }}-tarball'
+          path: ${{ runner.temp }}
+          merge-multiple: true
+
+      # This won't do anything if there are no artifacts to download (i.e. the base image didn't change)
+      # Retag these loaded images as latest, so that they get used as the base image instead of
+      # the one from DockerHub
+      - name: Load and tag base images
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        run: |
+          for file in ${{ runner.temp }}/*-${{ matrix.os }}.tar; do
+            docker load -i "$file" | sed -n 's/Loaded image: //p' | while read -r tagged_image; do
+              image="${tagged_image%:*}"
+              docker tag "$tagged_image" "$image:latest"
+            done
+          done
+          docker image ls -a
+
+      - name: Set docker tags and labels
+        id: meta
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: docker/metadata-action@v5
+        with:
+          context: git
+          images: |
+            ${{ inputs.image-prefix }}-${{ matrix.image }}
+          tags: |
+            type=raw,value=latest,enable={{is_default_branch}}
+            type=sha,prefix=,format=long
+
+      - name: Build ${{ (env.CAN_PUSH_IMAGE == 'true' && 'and push') || '' }}
+        id: build
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: docker/build-push-action@v6
+        continue-on-error: true
```

Comment:
I have observed some transient network issues sometimes in our CI (it fails, because something fails to download). I can remove this retry loop if wanted.

Reply:
What exactly fails to download? The artifact from the base image build jobs? Something else?

Reply:
Generally this is a failure in an apt package going OOM. See https://github.com/PrairieLearn/PrairieLearn/issues/11232. I have also seen d2 fail to download with a 403 occasionally. Maybe I can remove this until we observe another transient issue?

Reply:
Removed for now.